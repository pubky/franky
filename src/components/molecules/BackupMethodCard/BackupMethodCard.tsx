import * as Atoms from '@/atoms';
import * as Molecules from '@/molecules';

export const BackupMethodCard = () => {
  return (
    <Molecules.ContentCard
      image={{
        src: '/images/shield.png',
        alt: 'Shield',
        width: 192,
        height: 192,
        size: 'medium',
      }}
    >
      <Atoms.Container className="items-center gap-1 flex-row">
        <Atoms.Heading level={3} size="md">
          Choose backup method
        </Atoms.Heading>
        <Molecules.PopoverPublicKey />
      </Atoms.Container>
      <Atoms.Container className="flex-col gap-3">
        <Atoms.Typography className="text-secondary-foreground opacity-80 font-normal">
          Safely back up and store the secret seed for your pubky. Which backup method do you prefer? You can also
          choose to do this later.
        </Atoms.Typography>
      </Atoms.Container>
    </Molecules.ContentCard>
  );
};
