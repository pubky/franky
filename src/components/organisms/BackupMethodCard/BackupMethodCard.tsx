'use client';

import * as Atoms from '@/atoms';
import * as Molecules from '@/molecules';
import * as Stores from '@/core';
import * as Organisms from '@/organisms';

export const BackupMethodCard = () => {
  const { mnemonic } = Stores.useOnboardingStore();
  return (
    <Molecules.ContentCard
      image={{
        src: '/images/shield.png',
        alt: 'Shield',
        width: 192,
        height: 192,
      }}
    >
      <Atoms.Container className="items-center gap-1 flex-row">
        <Atoms.Heading level={2} size="lg">
          Choose backup method
        </Atoms.Heading>
        <Molecules.PopoverBackup />
      </Atoms.Container>
      <Atoms.Container className="max-w-[576px] mx-0">
        <Atoms.Typography size="sm" className="text-secondary-foreground opacity-80 font-medium text-base">
          Safely back up and store the secret seed for your pubky. Which backup method do you prefer? You can also
          choose to do this later.
        </Atoms.Typography>
        <Atoms.Container className="flex-row mt-6 gap-3 flex-wrap">
          <Molecules.DialogBackupPhrase />
          <Organisms.DialogBackupEncrypted />
          <Molecules.DialogExport mnemonic={mnemonic} />
        </Atoms.Container>
      </Atoms.Container>
    </Molecules.ContentCard>
  );
};
