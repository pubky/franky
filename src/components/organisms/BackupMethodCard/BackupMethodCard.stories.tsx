import type { Meta, StoryObj } from '@storybook/nextjs-vite';
import { BackupMethodCard } from './BackupMethodCard';

const meta = {
  title: 'Organisms/BackupMethodCard',
  component: BackupMethodCard,
  parameters: {
    layout: 'centered',
    docs: {
      description: {
        component:
          'A card component that allows users to choose their preferred backup method for their pubky. Displays three options: recovery phrase, encrypted file, and export to Pubky Ring. The export button text changes based on whether a mnemonic is available in the onboarding store.',
      },
    },
  },
  tags: ['autodocs'],
} satisfies Meta<typeof BackupMethodCard>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  parameters: {
    docs: {
      description: {
        story:
          'Default state of the BackupMethodCard component. The component will use the current state of the onboarding store to determine the export button text.',
      },
    },
  },
};
