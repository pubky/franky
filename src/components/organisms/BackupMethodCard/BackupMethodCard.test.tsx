import React from 'react';
import { describe, it, expect, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import { BackupMethodCard } from './BackupMethodCard';

// Mock onboarding store
const mockUseOnboardingStore = vi.fn();
vi.mock('@/core', () => ({
  useOnboardingStore: () => mockUseOnboardingStore(),
}));

// Mock atoms
vi.mock('@/atoms', () => ({
  Container: ({ children, className }: { children: React.ReactNode; className?: string }) => (
    <div data-testid="container" className={className}>
      {children}
    </div>
  ),
  Heading: ({
    children,
    level,
    size,
    className,
  }: {
    children: React.ReactNode;
    level: number;
    size?: string;
    className?: string;
  }) => (
    <div data-testid={`heading-${level}`} data-size={size} className={className}>
      {children}
    </div>
  ),
  Typography: ({ children, size, className }: { children: React.ReactNode; size?: string; className?: string }) => (
    <div data-testid="typography" data-size={size} className={className}>
      {children}
    </div>
  ),
  Dialog: vi.fn(({ children }: { children: React.ReactNode }) => <div data-testid="dialog">{children}</div>),
  DialogTrigger: vi.fn(({ children }: { children: React.ReactNode }) => (
    <div data-testid="dialog-trigger">{children}</div>
  )),
  DialogClose: vi.fn(({ children }: { children: React.ReactNode }) => <div data-testid="dialog-close">{children}</div>),
  DialogContent: ({ children, className }: { children: React.ReactNode; className?: string }) => (
    <div data-testid="dialog-content" className={className}>
      {children}
    </div>
  ),
  DialogHeader: vi.fn(() => <div data-testid="dialog-header" />),
  DialogTitle: vi.fn(() => <h2 data-testid="dialog-title" />),
  DialogDescription: vi.fn(() => <div data-testid="dialog-description" />),
  Button: ({
    children,
    variant,
    className,
    onClick,
  }: {
    children: React.ReactNode;
    variant?: string;
    className?: string;
    onClick?: () => void;
  }) => (
    <button data-testid="button" data-variant={variant} className={className} onClick={onClick}>
      {children}
    </button>
  ),
  Card: ({ children, className }: { children: React.ReactNode; className?: string }) => (
    <div data-testid="card" className={className}>
      {children}
    </div>
  ),
  Label: ({ children, htmlFor, className }: { children: React.ReactNode; htmlFor?: string; className?: string }) => (
    <label data-testid="label" htmlFor={htmlFor} className={className}>
      {children}
    </label>
  ),
  Input: ({
    type,
    id,
    placeholder,
    value,
    onChange,
    className,
  }: {
    type?: string;
    id?: string;
    placeholder?: string;
    value?: string;
    onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
    className?: string;
  }) => (
    <input
      data-testid="input"
      type={type}
      id={id}
      placeholder={placeholder}
      value={value}
      onChange={onChange}
      className={className}
    />
  ),
}));

// Mock libs
vi.mock('@/libs', () => ({
  ArrowLeft: ({ className }: { className?: string }) => (
    <div data-testid="arrow-left" className={className}>
      ArrowLeft
    </div>
  ),
  Download: ({ className }: { className?: string }) => (
    <div data-testid="download" className={className}>
      Download
    </div>
  ),
  Eye: ({ className }: { className?: string }) => (
    <div data-testid="eye" className={className}>
      Eye
    </div>
  ),
  EyeOff: ({ className }: { className?: string }) => (
    <div data-testid="eye-off" className={className}>
      EyeOff
    </div>
  ),
  Copy: ({ className }: { className?: string }) => (
    <div data-testid="copy" className={className}>
      Copy
    </div>
  ),
  Check: ({ className }: { className?: string }) => (
    <div data-testid="check" className={className}>
      Check
    </div>
  ),
  TriangleAlert: ({ className }: { className?: string }) => (
    <div data-testid="triangle-alert" className={className}>
      TriangleAlert
    </div>
  ),
  ShieldCheck: ({ className }: { className?: string }) => (
    <div data-testid="shield-check" className={className}>
      ShieldCheck
    </div>
  ),
  FileText: ({ className }: { className?: string }) => (
    <div data-testid="file-text" className={className}>
      FileText
    </div>
  ),
  Scan: ({ className }: { className?: string }) => (
    <div data-testid="scan" className={className}>
      Scan
    </div>
  ),
}));

// Mock molecules
vi.mock('@/molecules', () => ({
  ContentCard: ({
    children,
    image,
  }: {
    children: React.ReactNode;
    image?: { src: string; alt: string; width: number; height: number };
  }) => (
    <div data-testid="content-card" data-image-src={image?.src} data-image-alt={image?.alt}>
      {children}
    </div>
  ),
  PopoverBackup: () => <div data-testid="popover-backup">Backup Info</div>,
  DialogBackupPhrase: ({ children }: { children?: React.ReactNode }) => (
    <div data-testid="dialog-backup-phrase">{children || 'Backup Phrase'}</div>
  ),
  DialogExport: ({ mnemonic, children }: { mnemonic?: string; children?: React.ReactNode }) => (
    <div data-testid="dialog-export" data-mnemonic={mnemonic || ''}>
      {children || `Export ${mnemonic ? 'with mnemonic' : 'without mnemonic'}`}
    </div>
  ),
}));

// Mock organisms
vi.mock('@/organisms', () => ({
  DialogBackupEncrypted: ({ children }: { children?: React.ReactNode }) => (
    <div data-testid="dialog-backup-encrypted">{children || 'Backup Encrypted'}</div>
  ),
}));

describe('BackupMethodCard', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('renders with default mnemonic', () => {
    mockUseOnboardingStore.mockReturnValue({
      mnemonic: '',
    });

    render(<BackupMethodCard />);

    expect(screen.getByTestId('content-card')).toBeInTheDocument();
    expect(screen.getByTestId('heading-2')).toHaveTextContent('Choose backup method');
  });

  it('renders with mnemonic from store', () => {
    const testMnemonic = 'wood fox silver drive march fee palace flame earn door case almost';
    mockUseOnboardingStore.mockReturnValue({
      mnemonic: testMnemonic,
    });

    render(<BackupMethodCard />);

    const dialogExport = screen.getByTestId('dialog-export');
    expect(dialogExport).toHaveAttribute('data-mnemonic', testMnemonic);
    expect(dialogExport).toHaveTextContent('Export recovery phrase');
  });

  it('renders all backup options', () => {
    mockUseOnboardingStore.mockReturnValue({
      mnemonic: 'test mnemonic',
    });

    render(<BackupMethodCard />);

    expect(screen.getByTestId('dialog-backup-phrase')).toBeInTheDocument();
    expect(screen.getByTestId('dialog-backup-encrypted')).toBeInTheDocument();
    expect(screen.getByTestId('dialog-export')).toBeInTheDocument();
  });

  it('renders content card with shield image', () => {
    mockUseOnboardingStore.mockReturnValue({
      mnemonic: '',
    });

    render(<BackupMethodCard />);

    const contentCard = screen.getByTestId('content-card');
    expect(contentCard).toHaveAttribute('data-image-src', '/images/shield.png');
    expect(contentCard).toHaveAttribute('data-image-alt', 'Shield');
  });

  it('passes mnemonic correctly to DialogExport based on store state', () => {
    const testCases = [
      { mnemonic: '', expectedDisplay: 'Export to Pubky Ring' },
      { mnemonic: 'test phrase', expectedDisplay: 'Export recovery phrase' },
      {
        mnemonic: 'wood fox silver drive march fee palace flame earn door case almost',
        expectedDisplay: 'Export recovery phrase',
      },
    ];

    testCases.forEach(({ mnemonic, expectedDisplay }) => {
      mockUseOnboardingStore.mockReturnValue({ mnemonic });

      const { unmount } = render(<BackupMethodCard />);

      const dialogExport = screen.getByTestId('dialog-export');
      expect(dialogExport).toHaveAttribute('data-mnemonic', mnemonic);
      expect(dialogExport).toHaveTextContent(expectedDisplay);

      unmount();
    });
  });

  it('integrates correctly with onboarding store', () => {
    const testMnemonic = 'integration test mnemonic phrase';
    mockUseOnboardingStore.mockReturnValue({
      mnemonic: testMnemonic,
      pubky: 'test-public-key',
      secretKey: 'test-secret-key',
      hasHydrated: true,
    });

    render(<BackupMethodCard />);

    // Should call the store hook
    expect(mockUseOnboardingStore).toHaveBeenCalled();

    // Should pass the mnemonic to DialogExport
    const dialogExport = screen.getByTestId('dialog-export');
    expect(dialogExport).toHaveAttribute('data-mnemonic', testMnemonic);
  });

  describe('organism behavior', () => {
    it('handles undefined mnemonic gracefully', () => {
      mockUseOnboardingStore.mockReturnValue({
        mnemonic: undefined,
      });

      render(<BackupMethodCard />);

      const dialogExport = screen.getByTestId('dialog-export');
      // When mnemonic is undefined, it should be treated as empty string
      expect(dialogExport).toHaveAttribute('data-mnemonic', '');
      expect(dialogExport).toHaveTextContent('Export to Pubky Ring');
    });

    it('maintains component structure regardless of store state', () => {
      const storeStates = [{ mnemonic: '' }, { mnemonic: 'test' }, { mnemonic: undefined }, { mnemonic: null }];

      storeStates.forEach((storeState) => {
        mockUseOnboardingStore.mockReturnValue(storeState);

        const { unmount } = render(<BackupMethodCard />);

        // All main components should always be present
        expect(screen.getByTestId('content-card')).toBeInTheDocument();
        expect(screen.getByTestId('heading-2')).toBeInTheDocument();
        expect(screen.getByTestId('popover-backup')).toBeInTheDocument();
        expect(screen.getByTestId('typography')).toBeInTheDocument();
        expect(screen.getByTestId('dialog-backup-phrase')).toBeInTheDocument();
        expect(screen.getByTestId('dialog-backup-encrypted')).toBeInTheDocument();
        expect(screen.getByTestId('dialog-export')).toBeInTheDocument();

        unmount();
      });
    });
  });
});

describe('BackupMethodCard - Snapshots', () => {
  it('matches snapshot for default BackupMethodCard', () => {
    mockUseOnboardingStore.mockReturnValue({
      mnemonic: '',
    });

    const { container } = render(<BackupMethodCard />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
