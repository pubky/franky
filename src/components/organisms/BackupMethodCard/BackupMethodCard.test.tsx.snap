// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BackupMethodCard - Snapshots > matches snapshot for default BackupMethodCard 1`] = `
<div
  data-image-alt="Shield"
  data-image-src="/images/shield.png"
  data-testid="content-card"
>
  <div
    class="items-center gap-1 flex-row"
    data-testid="container"
  >
    <div
      data-size="lg"
      data-testid="heading-2"
    >
      Choose backup method
    </div>
    <div
      data-testid="popover-backup"
    >
      Backup Info
    </div>
  </div>
  <div
    class="max-w-[576px] mx-0"
    data-testid="container"
  >
    <div
      class="text-secondary-foreground opacity-80 font-medium text-base"
      data-size="sm"
      data-testid="typography"
    >
      Safely back up and store the secret seed for your pubky. Which backup method do you prefer? You can also choose to do this later.
    </div>
    <div
      class="flex-row mt-6 gap-3 flex-wrap"
      data-testid="container"
    >
      <div
        data-testid="dialog-backup-phrase"
      >
        <button
          class="gap-2"
          data-testid="button"
          data-variant="secondary"
        >
          <div
            class="h-4 w-4"
            data-testid="file-text"
          >
            FileText
          </div>
          <span>
            Recovery phrase
          </span>
        </button>
      </div>
      <div
        data-testid="dialog-backup-encrypted"
      >
        <button
          class="gap-2"
          data-testid="button"
          data-variant="secondary"
        >
          <div
            class="h-4 w-4"
            data-testid="file-text"
          >
            FileText
          </div>
          <span>
            Encrypted file
          </span>
        </button>
      </div>
      <div
        data-mnemonic=""
        data-testid="dialog-export"
      >
        <button
          class="gap-2"
          data-testid="button"
        >
          <div
            class="h-4 w-4"
            data-testid="scan"
          >
            Scan
          </div>
          <span>
            Export to Pubky Ring
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
